pkg/kubelet/deviceplugin/manager.go

```diff
@@ -212,21 +213,38 @@ func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
 		return
 	}
 
-	stream, err := e.list()
-	if err != nil {
-		glog.Errorf("Failed to List devices for plugin %v: %v", r.ResourceName, err)
-		return
-	}
+	closeChan := make(chan bool)
+	go e.run(closeChan)
 
 	go func() {
-		e.listAndWatch(stream)
-
-		m.mutex.Lock()
-		if old, ok := m.Endpoints[r.ResourceName]; ok && old == e {
-			delete(m.Endpoints, r.ResourceName)
+		// TODO: Consider making timeout value configurable.
+		t := time.NewTimer(5 * time.Second)
+		defer t.Stop()
+
+		for {
+			select {
+			case <-t.C:
+				m.mutex.Lock()
+				if old, ok := m.Endpoints[r.ResourceName]; ok && old == e {
+					if len(e.devices) == 0 {
+						glog.V(2).Infof("No device registered after 5 seconds, unregistered endpoint %v", e)
+						delete(m.Endpoints, r.ResourceName)
+						e.stop()
+						m.mutex.Unlock()
```

Comment:
If you defer the mutex Unlock right after the lock you won't have to unlock it twice.
It's also usually more idiomatic :)

Reply:
good catch

Reply:
good catch